src/devices/bus/nabupc/adapter.cpp

```diff
@@ -0,0 +1,375 @@
+	if (err)
+		return err;
+
+	memset(current.data, 0, 991);
+	current.segment_id[0] = (segment_id & 0xFF0000) >> 16;
+	current.segment_id[1] = (segment_id & 0x00FF00) >> 8;
+	current.segment_id[2] = (segment_id & 0xF000FF);
+	current.owner         = 0x01;
+	current.tier[0]       = 0x7f;
+	current.tier[1]       = 0xff;
+	current.tier[2]       = 0xff;
+	current.tier[3]       = 0xff;
+	current.mbytes[0]     = 0x7f;
+	current.mbytes[1]     = 0x80;
+	err = fd->read(current.data, offset, 991, actual);
+	do {
+		crc = 0xffff;
+		if (err) {
+			return err;
+		}
+		if (actual > 0) {
+			current.packet_number = npak;
+			current.pak_number[0] = npak;
+			current.pak_number[1] = 0;
+			current.type = 0x20;
+			if (offset == 0)
+				current.type |= 0x81;
+			if (actual < 991)
+				current.type |= 0x10;
+			current.offset[0] = ((offset) >> 8) & 0xFF;
+			current.offset[1] = offset & 0xFF;
+			for (int i = 0; i < 1007; ++i) {
+				crc = update_crc(crc, ((char *)&current)[i]);
+			}
+			crc ^= 0xffff;
+			current.crc[0] = (crc >> 8) & 0xFF;
+			current.crc[1] = crc & 0xFF;
+			pak_list.push_back(current);
+			offset = (++npak * 991);
+			memset(current.data, 0, 991);
+			err = fd->read(current.data, offset, 991, actual);
+		}
+	} while(actual > 0);
+
+	if (err) {
+		return err;
+	}
+
+	return std::error_condition();
+}
+
+const segment_file::pak& segment_file::operator[](const int index) const
+{
+	assert(index >= 0 && index < size());
+
+	return pak_list[index];
+}
+
+// crc16 calculation
+uint16_t segment_file::update_crc(uint16_t crc, uint8_t data)
+{
+	uint8_t bc;
+
+	bc = (crc >> 8) ^ data;
+
+	crc <<= 8;
+	crc ^= crc_table[bc];
+
+	return crc;
+}
+
+//**************************************************************************
+//  INPUT PORTS
+//**************************************************************************
+
+// CONFIG
+static INPUT_PORTS_START( nabu_network_adapter )
+	PORT_START("CONFIG")
+	PORT_CONFNAME(0x01, 0x00, "Prompt for channel?")
+	PORT_CONFSETTING(0x01, "Yes")
+	PORT_CONFSETTING(0x00, "No")
+INPUT_PORTS_END
+
+//**************************************************************************
+//  DEVICE INITIALIZATION
+//**************************************************************************
+
+network_adapter::network_adapter(machine_config const &mconfig, char const *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, NABUPC_NETWORK_ADAPTER, tag, owner, clock)
+	, device_buffered_serial_interface(mconfig, *this)
+	, device_rs232_port_interface(mconfig, *this)
+	, m_config(*this, "CONFIG")
+	, m_channel(0)
+	, m_packet(0)
+	, m_segment(0)
+	, m_state(State::START)
+	, m_substate(0)
+	, m_segment_timer(nullptr)
+	{
+}
+
+void network_adapter::device_start()
+{
+	m_segment_timer = timer_alloc(FUNC(network_adapter::segment_tick), this);
+
+	save_item(NAME(m_state));
+	save_item(NAME(m_substate));
+	save_item(NAME(m_channel));
+	save_item(NAME(m_packet));
+	save_item(NAME(m_segment));
```

Comment:
Yeah i was not sure how to handle saving a complex data type like an internal class. Didn't know about the postload handler, so I'll try and add one. Though since the segment being loaded is external, I'm not sure how this would behave in the case where the segment file was deleted or changed since the initial save state was taken. On the other hand the state of the cache is really only going to be an issue if they saved during a segment load process.